vllm/worker/cache_engine.py

```diff
@@ -43,95 +42,49 @@ def __init__(
         self.gpu_cache = self.allocate_gpu_cache()
         self.cpu_cache = self.allocate_cpu_cache()
 
-    def get_key_block_shape(self) -> Tuple[int, int, int, int]:
-        element_size = torch.tensor([], dtype=self.dtype).element_size()
-        x = 16 // element_size
-        return (
-            self.num_heads,
-            self.head_size // x,
-            self.block_size,
-            x,
-        )
-
-    def get_value_block_shape(self) -> Tuple[int, int, int]:
-        return (
-            self.num_heads,
-            self.head_size,
-            self.block_size,
-        )
-
-    def allocate_gpu_cache(self) -> List[KVCache]:
-        gpu_cache: List[KVCache] = []
-        key_block_shape = self.get_key_block_shape()
-        value_block_shape = self.get_value_block_shape()
-        for _ in range(self.num_layers):
-            key_blocks = torch.empty(
-                size=(self.num_gpu_blocks, *key_block_shape),
-                dtype=self.dtype,
-                device="cuda",
-            )
-            value_blocks = torch.empty(
-                size=(self.num_gpu_blocks, *value_block_shape),
-                dtype=self.dtype,
-                device="cuda",
-            )
-            gpu_cache.append((key_blocks, value_blocks))
+    def allocate_gpu_cache(self) -> List[torch.Tensor]:
+        kv_cache_shape = self.attn_backend.get_kv_cache_shape(
+            self.num_gpu_blocks, self.block_size, self.num_heads,
+            self.head_size)
+        gpu_cache = [
+            torch.empty(kv_cache_shape, dtype=self.dtype, device="cuda")
+            for _ in range(self.num_layers)
+        ]
         return gpu_cache
 
-    def allocate_cpu_cache(self) -> List[KVCache]:
-        cpu_cache: List[KVCache] = []
-        key_block_shape = self.get_key_block_shape()
-        value_block_shape = self.get_value_block_shape()
+    def allocate_cpu_cache(self) -> List[torch.Tensor]:
         pin_memory = is_pin_memory_available()
+        kv_cache_shape = self.attn_backend.get_kv_cache_shape(
+            self.num_cpu_blocks, self.block_size, self.num_heads,
+            self.head_size)
+        cpu_cache: List[torch.Tensor] = []
         for _ in range(self.num_layers):
-            key_blocks = torch.empty(
-                size=(self.num_cpu_blocks, *key_block_shape),
-                dtype=self.dtype,
-                pin_memory=pin_memory,
-                device="cpu",
-            )
-            value_blocks = torch.empty(
-                size=(self.num_cpu_blocks, *value_block_shape),
-                dtype=self.dtype,
-                pin_memory=pin_memory,
-                device="cpu",
-            )
-            cpu_cache.append((key_blocks, value_blocks))
+            cpu_cache.append(
+                torch.empty(kv_cache_shape,
+                            dtype=self.dtype,
+                            pin_memory=pin_memory,
+                            device="cpu"))
         return cpu_cache
```

Comment:
These two functions seems can be merged into one?

Reply:
Good point! Fixed.